IRCv3.3 Message Tags

jwheare opened the pull request:
This specification adds a new capability for general message tags support and
a prefix for expressing client-to-client tags.

This PR replaces #236

kylef commented on message-tags-3.3.md:
```diff
@@ -0,0 +1,71 @@
+MAY be forwarded on other events.
+
+The expected client behaviour of individual client-to-client tags SHOULD be defined
+in separate specifications, in the same way as server-initiated tags.
```
I'd love to see some clarification in the spec if tag keys themselves within a message are unique or can be used multiple times.

Whether the following is valid/expected or not:

```
@+icon=https://example.com/favicon.png;+icon=something :url_bot!bot@example.com PRIVMSG #channel :Example.com: A News Story
```

DarthGandalf commented on message-tags-3.3.md:
```diff
@@ -0,0 +1,75 @@
+[IRCv3.2 tags][irctags].
+
+Client-to-client tags MUST be relayed on `PRIVMSG` and `NOTICE` events, and
+MAY be forwarded on other events.
```
Relayed vs forwarded. Let's use a single term for single meaning.

DanielOaks commented on core/message-tags-3.3.md:
```diff
@@ -0,0 +1,79 @@
+Client-to-client tags are client-initiated tags that servers MUST attach as-is
+to any relevant event relayed to other clients. A client-to-client tag is prefixed
+with a plus sign (`+`) and otherwise conforms to the format specified in
+[IRCv3.2 tags][irctags].
```
`[IRCv3.2 tags](./message-tags-3.2.html).`

DanielOaks commented:
Looks good to me, +1 on merging.

jwheare commented:
* Added new sections for client implementation and moderation considerations.
* Rewrote the motivation section.
* Renamed "client-to-client" to "client-only", less of a mouthful.

jwheare commented:
Last call for comments.

grawity commented:
Remind me, how does a _client_ know that it can send tags? Does the new cap work in both directions?

Looks good overall.

jwheare commented:
> Remind me, how does a client know that it can send tags? Does the new cap work in both directions?

Yes. I'll clarify.

SaberUK commented on core/message-tags-3.3.md:
```diff
@@ -0,0 +1,100 @@
+
+Individual tag keys MUST only be used a maximum of once per message. Clients
+receiving messages with more than one occurrence of a tag key SHOULD discard all
+but the final occurrence.
```
Perhaps we should add a reminder around here that people should use vendor prefixes for proprietary features.

jwheare replied:
Done.

SaberUK commented:
LGTM. :shipit:

jwheare commented:
Squashed

dequis commented:
LGTM

attilamolnar commented:
LGTM

SaberUK commented:
LGTM :shipit:
